Multilevel logging extra APIs

jzuniga-amd opened the pull request:
I would like to separate the logger creation/destruction from the logger manipulation (initialization and cleanup). For that I added two new APIs: rocsolver_create_logger and rocsolver_destroy_logger, and modified a bit the behavior of rocsolver_logging_cleanup and rocsolver_logging_initialize.

An application can be logged by simply having rocsolver_create_logger and rocsolver_destroy_logger at the beginning and the end of the code, and using the environment variables. Additionally, we can have more advanced/versatile logging by using the other two APIs within the code: the logging mode can be changed on the fly, or stopped for some parts of the code, we can print intermediate profile results, etc....

tfalders reviewed (approved):
LGTM! Feel free to merge it in and I'll update the relevant documentation in the user guide.

tfalders commented on rocsolver/library/src/rocsolver_logger.cpp:
```diff
@@ -150,12 +182,38 @@ rocblas_status rocsolver_logging_cleanup()
-    delete logger;
-    logger = nullptr;
+    // reset to no logging and clear profile info
+    logger->max_levels = 1;
+    logger->layer_mode = rocblas_layer_mode_none;
```
It just occurred to me that the values provided using environment variables will only be used upon first creating the logger. Should we default back to these values when doing the cleanup?

jzuniga-amd replied:
I rather go back to the non-logging default. This way cleanup is also used as a way to suspend logging which can be restarted with the loggin_initialize. As the static logger object is not deleted we can even pass an option to clean or not the profile unordered map; this way a user can suspend profiling in a section of its application code, and re-activate it again later...

jzuniga-amd commented:
I added the clean_profile option. 
I also updated the examples with the new APIs but, as mentioned in the main PR, you may want to create a separate example sample to illustrate the whole logging functionality (it can be done in a future PR too)

jzuniga-amd commented:
You have to merge it if you agree...
I have no write access :)
